### PR TITLE
largest divisible subset

### PR DESCRIPTION
**Pull request by pa-kh039**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by pa-kh039** [comment]
@tianyizheng02 @cclauss  please review   (PR for hacktoberfest 2023)

**Comment by cclauss** [comment]
When would someone want to use this algorithm?

**Comment by pa-kh039** [comment]
> When would someone want to use this algorithm?


There are several possible use cases for this algorithm. To enlist a few:
1. When managing a portfolio of financial assets, you might want to find the largest subset of assets where each asset's price is a multiple of another, which can help with diversification and risk management.
2. In data analysis, you may encounter situations where you want to group data points into clusters based on some criteria, and this problem can be used to find the largest subset of data points that are related according to specific rules.

So, with some changes in it , this algorithm can be utilised at several places.

File: dynamic_programming/largest_divisible_subset.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    Algorithm to find the biggest subset
+    in the given array such that for any
+    2 elements x and y in the subset,
+    either x divides y or y divides x
+    >>> largest_divisible_subset([1,16,7,8,4])
+    [16, 8, 4, 1]
+    >>> largest_divisible_subset([1,2,3])
+    [2, 1]
+    >>> largest_divisible_subset([1, 2, 4, 8])
+    [8, 4, 2, 1]
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:15]: Let's add some test cases.
```suggestion
    >>> largest_divisible_subset([1, 16, 7, 8, 4])
    [16, 8, 4, 1]
    >>> largest_divisible_subset([1, 2, 3])
    [2, 1]
    >>> largest_divisible_subset([-1, -2, -3])
    [-3]
    >>> largest_divisible_subset([1, 2, 4, 8])
    [8, 4, 2, 1]
    >>> largest_divisible_subset((1, 2, 4, 8))
    [8, 4, 2, 1]
    >>> largest_divisible_subset([1, 1, 1])
    [1, 1, 1]
    >>> largest_divisible_subset([0, 0, 0])
    [0, 0, 0]
    >>> largest_divisible_subset([-1, -1, -1])
    [-1, -1, -1]
    >>> largest_divisible_subset([])
    []
```

File: dynamic_programming/largest_divisible_subset.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    array.sort()
+
+    # Initialize memo and hash arrays with 1s
+    memo = [1] * array_size
+    hash_array = list(range(array_size))
+
+    # Iterate through the array
+    for i in range(array_size):
+        for prev_index in range(i):
+            if array[i] % array[prev_index] == 0 and 1 + memo[prev_index] > memo[i]:
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:31]:
```suggestion
    for i, item in enumerate(array):
        for prev_index in range(i):
            if item % array[prev_index] == 0 and 1 + memo[prev_index] > memo[i]:
```

File: dynamic_programming/largest_divisible_subset.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                memo[i] = 1 + memo[prev_index]
+                hash_array[i] = prev_index
+
+    ans = -1
+    last_index = -1
+
+    # Find the maximum length and its corresponding index
+    for i in range(array_size):
+        if memo[i] > ans:
+            ans = memo[i]
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:53]:
```suggestion
    for i, memo_item in enumerate(memo):
        if memo_item > ans:
            ans = memo_item
```

File: dynamic_programming/largest_divisible_subset.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+
+def largest_divisible_subset(array: list[int]) -> list[int]:
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:4]: `array` is not really a self-documenting name.  This is an array of what?

File: dynamic_programming/largest_divisible_subset.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+if __name__ == "__main__":
+    from doctest import testmod
+
+    testmod()
+
+    array = [1, 16, 7, 8, 4]
+
+    answer = largest_divisible_subset(array)
+
+    print("The longest divisible subset elements are:", answer)
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:74]:
```suggestion
    items = [1, 16, 7, 8, 4]
    print(f"The longest divisible subset of {items} is {largest_divisible_subset(items)}.")

    answer = largest_divisible_subset(array)

    print("The longest divisible subset elements are:", answer)
```

**Comment by pa-kh039** [comment]
@cclauss Did all the changes suggested by you and all the associated changes.

File: dynamic_programming/largest_divisible_subset.py [hunk 6]
```diff
@@ -15,7 +15,7 @@ def largest_divisible_subset(array: list[int]) -> list[int]:
     [-3]
     >>> largest_divisible_subset([1, 2, 4, 8])
     [8, 4, 2, 1]
-    >>> largest_divisible_subset((1, 2, 4, 8))
+    >>> largest_divisible_subset([1, 2, 4, 8])
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:18]: No.  We want to test tuples.

File: dynamic_programming/largest_divisible_subset.py [hunk 7]
```diff
@@ -71,7 +75,11 @@ def largest_divisible_subset(array: list[int]) -> list[int]:
     print(
         f"The longest divisible subset of {items} is {largest_divisible_subset(items)}."
     )
-
-    answer = largest_divisible_subset(array)
-
-    print("The longest divisible subset elements are:", answer)
+    if len(items) == 0:
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py:76]: Avoid the function call overhead and deal gracefully with being passed None.
```suggestion
    if not items:
```

File: dynamic_programming/largest_divisible_subset.py [hunk 8]
```diff
@@ -71,7 +75,11 @@ def largest_divisible_subset(array: list[int]) -> list[int]:
...
+        print("No items to form subset!")
+    else:
+        answer = largest_divisible_subset(items)
+        if len(answer) == 0:
+            print("No subset found")
+        else:
+            print("The longest divisible subset elements are:", answer)
```
Comment (cclauss) [inline dynamic_programming/largest_divisible_subset.py]:
```suggestion
```

**Comment by pa-kh039** [comment]
@cclauss Would you like to suggest any more changes in the latest version?

**Review by cclauss — APPROVED** [review]

**Pull request merged**